ramips: Add support for ZyXEL Keenetic Viva

13hakta opened the pull request:
Add support for ZyXEL Keenetic Viva.
Add build option for RTL8367R/VB switches.

Signed-off-by: Vitaly Chekryzhev 13hakta@gmail.com

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -242,6 +242,15 @@ ramips_setup_interfaces()
 			"1:lan" "2:lan" "0:wan" "6@eth0"
 		;;
+	kng_rc)
+		ucidef_set_interfaces_lan_wan "eth0.1" "eth0.2"
```
Shouldn't be required. The lan/wan interfaces are setup by ```uci_add_switch```.

13hakta replied:
ok

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -242,6 +242,15 @@ ramips_setup_interfaces()
+			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "7t@eth0"
+
+		ucidef_add_switch_vlan "switch1" "1" "0 1 2 3 7t"
+		ucidef_add_switch_vlan "switch1" "2" "4 7t"
```
Shouldn't be required. The vlans are setup by ```uci_add_switch```.

mkresin commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -232,6 +232,9 @@ ramips_board_detect() {
 		name="jhr-n926r"
 		;;
+	*"ZyXEL Keenetic Viva")
+		name="kng_rc"
```
I'm not that satisfied with the name kng_rc. I don't see any relation to the full board name.

13hakta replied:
Address this to NDM. Reference device table here: https://github.com/ndmsystems/packages/wiki/NDM-Systems-Buildroot#configuration

mkresin replied:
I have to admit I fail to see how ndmsystems is related to ZyXEL.

Nevertheless, just because $vendor uses a shitty shortname doesn't mean we have to use the same shitty shortname. Or do I miss the point here?

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -358,6 +358,11 @@ youku-yk1)
 	set_usb_led "$board:blue:usb"
 	;;
+kng_rc)
+	ucidef_set_led_default "power" "power" "$board:green:power" "1"
```
Shouldn't be required. The power led is set to permanent on in diag.sh => set_state => done

mkresin commented on target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom:
```diff
@@ -43,6 +43,9 @@ board=$(ramips_board_name)
 	case $board in
+	kng_rc)
+		rt2x00_eeprom_extract "factory" 0 512
+	;;
```
please drop this. it is redundant to

```
&wmac {
	ralink,mtd-eeprom = <&factory 0>;
};
```

in your dts file.

13hakta replied:
ok

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,183 @@
+/dts-v1/;
+
+#include "mt7620a.dtsi"
+
```
Please include <[dt-bindings/input/input.h](http://lxr.free-electrons.com/source/include/dt-bindings/input/linux-event-codes.h?v=4.4)> and  <[dt-bindings/gpio/gpio.h](http://lxr.free-electrons.com/source/include/dt-bindings/gpio/gpio.h?v=4.4)> here as well.

Use the constants GPIO_ACTIVE_LOW and GPIO_ACTIVE_HIGH afterwards in stead of 1 and 0 in the gpio parameters.

Use the Key constants instead of the hex values in linux,code.

13hakta replied:
ok, done

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,183 @@
+		reset {
+			label = "reset";
+			gpios = <&gpio0 1 1>;
+			linux,code = <0x101>;
```
use KEY_RESTART here

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,183 @@
+		wps {
+			label = "wps";
+			gpios = <&gpio0 2 1>;
+			linux,code = <0x100>;
```
use KEY_WPS_BUTTON here

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,183 @@
+		fn {
+			label = "fn";
+			gpios = <&gpio0 9 1>;
+			linux,code = <0x102>;
```
... BTN_0

mkresin commented:
> Address this to NDM. Reference device table here: https://github.com/ndmsystems/packages/wiki/NDM-Systems-Buildroot#configuration


I have to admit I fail to see how ndmsystems is related to ZyXEL.

Nevertheless, just because $vendor uses a shitty shortname doesn't mean we have to use the same shitty shortname. Or do I miss the point here?

mkresin commented:
Please create an extra commit for the RTL8367R/VB switch support added to package/kernel/linux/modules/netdevices.mk. Add add it to this PR.

You know, one feature per commit.

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -358,6 +358,10 @@ youku-yk1)
 	set_wifi_led "$board:blue:air"
 	set_usb_led "$board:blue:usb"
 	;;
+kng_rc)
```
Please take care of the alphabetical ordering used in this file.

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -242,6 +242,10 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"1:lan" "2:lan" "0:wan" "6@eth0"
 		;;
+	kng_rc)
```
Please take care of the alphabetical ordering used in this file.

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -382,6 +386,9 @@ ramips_setup_macs()
 		wan_mac="$(grep -m1 mac= "/dev/mtd${index}" | cut -d= -f2)"
 		lan_mac=$wan_mac
 		;;
+	kng_rc)
```
Please take care of the alphabetical ordering used in this file.

mkresin commented on target/linux/ramips/base-files/etc/diag.sh:
```diff
@@ -212,6 +212,7 @@ get_status_led() {
 	wizfi630a)
 		status_led="$board::run"
 		;;
+	kng_rc|\
```
move the whole block up to keep alphabetical ordering

13hakta replied:
ok

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -242,6 +242,10 @@ ramips_setup_interfaces()
 			"1:lan" "2:lan" "0:wan" "6@eth0"
 		;;
+	kng_rc)
+		ucidef_add_switch "switch1" \
```
Shouldn't it be switch0?

13hakta replied:
No. There are two of them in device. Switch0 - unmanaged mt7530 and managed switch1 rtl8367rb.

mkresin commented:
Would it be possible that you squash your commits so that we have three commits in the end:

- kernel: add rtl8367b package
- ramips: add support for ZyXEL Keenetic Viva
- ramips: add MT7620 MIB support for switch and port.

The alphabetical order in ramips_setup_macs() (target/linux/ramips/base-files/etc/board.d/02_network) is still wrong. Beside the fact that I'm still not  happy with the short board name the PR looks good to me.

mkresin commented:
Hey Vitaly, short follow-up.

Over at openwrt some have send a PR which adds support for a few other Zyxel Keenetic boards (https://github.com/openwrt/openwrt/pull/65). This PR includes an tool to create a Zyxel image header (factory image), which might allow to upgrade via the webinterface firmware upgrade page.

Is something similar required for this board as well? How does the switch to LEDE happens on this board?

13hakta commented:
Updated. Lately I flashed image with TFTP server. But you are right, it's better to sign image with zyimage, that will allow burn image with WebUI. Also added this util to basic zyxel images support.

mkresin commented:
I've:

- squashed your commits
- created an extra commit for zyxel image tool
- moved the zyxel image tool to the firmware utils package
- removed the zyxel image binary
- prefixed the commit titles
- merged your MIB patch with the patch which adds the code you are changing

and pulled the PR into my staging tree: https://git.lede-project.org/?p=lede/mkresin/staging.git;a=summary.

Please take care of doing the outlined stuff by yourself with your next submission.

@13hakta Please compile and runtime test an image from my staging tree and confirm that my changes of your PR do not break anything!

@blogic Would you please ACK the MT7620 switch changes! I've applied them to the existing add MT7620 support patch since it doesn't look like that an upstream merge of this patch is pending.

blogic commented:
@mkresin ack

13hakta commented:
Thanks for help, that was first commit for experience to this project. Confirm build from your staging tree.

mkresin commented:
> Thanks for help, that was first commit for experience to this project. Confirm build from your staging tree.
> 
But does the build image works on the router? The MIB changes as well?

13hakta commented:
@mkresin MIB changes are ok. Recent changes make device work initially. More compatibility commits will follow.

mkresin commented:
Okay I will try it one more time (third time the same question!). If you don't answer my question, I will drop this PR from my staging tree.

Does an image build from my staging tree works as expected on the router? Please paste a bootlog here to confirm your answer.

> Recent changes make device work initially. More compatibility commits will follow.

What does work and what does not work. What kind of compatibility commit will follow? I'm not really interested to commit a half working board support.

13hakta commented:
Image build works as expected. Partition sizes, LED, buttons, WiFi, internal switch are working. External switch driver is not specific for ramips architecture (generic).

Here is bootlog:
`[    0.000000] Linux version 4.4.30 (13hakta@virtual-machine) (gcc version 5.4.0 (LEDE GCC 5.4.0 r2176) ) #0 Sun Nov 14 07:39:08 2016
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:4
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is ZyXEL Keenetic Viva
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000] free_area_init_node: node 0, pgdat 8038a3b0, node_mem_map 81000000
[    0.000000]   Normal zone: 256 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=0001e000
[    0.000000] Readback ErrCtl register=0001e000
[    0.000000] Memory: 125736K/131072K available (2895K kernel code, 137K rwdata, 696K rodata, 168K init, 203K bss, 5336K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: enable autosleep mode
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000012] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015448] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.080605] pid_max: default: 32768 minimum: 301
[    0.089927] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.102941] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.123875] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.143529] pinctrl core: initialized pinctrl subsystem
[    0.154500] NET: Registered protocol family 16
[    0.166315] rt2880-pinmux pinctrl: try to register 73 pins ...
[    0.166342] pinctrl core: registered pin 0 (io0) on rt2880-pinmux
[    0.166374] pinctrl core: registered pin 1 (io1) on rt2880-pinmux
[    0.166388] pinctrl core: registered pin 2 (io2) on rt2880-pinmux
[    0.166401] pinctrl core: registered pin 3 (io3) on rt2880-pinmux
[    0.166415] pinctrl core: registered pin 4 (io4) on rt2880-pinmux
[    0.166428] pinctrl core: registered pin 5 (io5) on rt2880-pinmux
[    0.166442] pinctrl core: registered pin 6 (io6) on rt2880-pinmux
[    0.166455] pinctrl core: registered pin 7 (io7) on rt2880-pinmux
[    0.166467] pinctrl core: registered pin 8 (io8) on rt2880-pinmux
[    0.166480] pinctrl core: registered pin 9 (io9) on rt2880-pinmux
[    0.166493] pinctrl core: registered pin 10 (io10) on rt2880-pinmux
[    0.166507] pinctrl core: registered pin 11 (io11) on rt2880-pinmux
[    0.166520] pinctrl core: registered pin 12 (io12) on rt2880-pinmux
[    0.166534] pinctrl core: registered pin 13 (io13) on rt2880-pinmux
[    0.166549] pinctrl core: registered pin 14 (io14) on rt2880-pinmux
[    0.166561] pinctrl core: registered pin 15 (io15) on rt2880-pinmux
[    0.166579] pinctrl core: registered pin 16 (io16) on rt2880-pinmux
[    0.166594] pinctrl core: registered pin 17 (io17) on rt2880-pinmux
[    0.166608] pinctrl core: registered pin 18 (io18) on rt2880-pinmux
[    0.166620] pinctrl core: registered pin 19 (io19) on rt2880-pinmux
[    0.166634] pinctrl core: registered pin 20 (io20) on rt2880-pinmux
[    0.166648] pinctrl core: registered pin 21 (io21) on rt2880-pinmux
[    0.166661] pinctrl core: registered pin 22 (io22) on rt2880-pinmux
[    0.166673] pinctrl core: registered pin 23 (io23) on rt2880-pinmux
[    0.166686] pinctrl core: registered pin 24 (io24) on rt2880-pinmux
[    0.166699] pinctrl core: registered pin 25 (io25) on rt2880-pinmux
[    0.166712] pinctrl core: registered pin 26 (io26) on rt2880-pinmux
[    0.166724] pinctrl core: registered pin 27 (io27) on rt2880-pinmux
[    0.166737] pinctrl core: registered pin 28 (io28) on rt2880-pinmux
[    0.166750] pinctrl core: registered pin 29 (io29) on rt2880-pinmux
[    0.166764] pinctrl core: registered pin 30 (io30) on rt2880-pinmux
[    0.166777] pinctrl core: registered pin 31 (io31) on rt2880-pinmux
[    0.166789] pinctrl core: registered pin 32 (io32) on rt2880-pinmux
[    0.166802] pinctrl core: registered pin 33 (io33) on rt2880-pinmux
[    0.166815] pinctrl core: registered pin 34 (io34) on rt2880-pinmux
[    0.166828] pinctrl core: registered pin 35 (io35) on rt2880-pinmux
[    0.166840] pinctrl core: registered pin 36 (io36) on rt2880-pinmux
[    0.166853] pinctrl core: registered pin 37 (io37) on rt2880-pinmux
[    0.166868] pinctrl core: registered pin 38 (io38) on rt2880-pinmux
[    0.166880] pinctrl core: registered pin 39 (io39) on rt2880-pinmux
[    0.166892] pinctrl core: registered pin 40 (io40) on rt2880-pinmux
[    0.166906] pinctrl core: registered pin 41 (io41) on rt2880-pinmux
[    0.166919] pinctrl core: registered pin 42 (io42) on rt2880-pinmux
[    0.166931] pinctrl core: registered pin 43 (io43) on rt2880-pinmux
[    0.166943] pinctrl core: registered pin 44 (io44) on rt2880-pinmux
[    0.166956] pinctrl core: registered pin 45 (io45) on rt2880-pinmux
[    0.166970] pinctrl core: registered pin 46 (io46) on rt2880-pinmux
[    0.166982] pinctrl core: registered pin 47 (io47) on rt2880-pinmux
[    0.167001] pinctrl core: registered pin 48 (io48) on rt2880-pinmux
[    0.167014] pinctrl core: registered pin 49 (io49) on rt2880-pinmux
[    0.167027] pinctrl core: registered pin 50 (io50) on rt2880-pinmux
[    0.167039] pinctrl core: registered pin 51 (io51) on rt2880-pinmux
[    0.167052] pinctrl core: registered pin 52 (io52) on rt2880-pinmux
[    0.167065] pinctrl core: registered pin 53 (io53) on rt2880-pinmux
[    0.167078] pinctrl core: registered pin 54 (io54) on rt2880-pinmux
[    0.167091] pinctrl core: registered pin 55 (io55) on rt2880-pinmux
[    0.167104] pinctrl core: registered pin 56 (io56) on rt2880-pinmux
[    0.167117] pinctrl core: registered pin 57 (io57) on rt2880-pinmux
[    0.167131] pinctrl core: registered pin 58 (io58) on rt2880-pinmux
[    0.167143] pinctrl core: registered pin 59 (io59) on rt2880-pinmux
[    0.167155] pinctrl core: registered pin 60 (io60) on rt2880-pinmux
[    0.167168] pinctrl core: registered pin 61 (io61) on rt2880-pinmux
[    0.167181] pinctrl core: registered pin 62 (io62) on rt2880-pinmux
[    0.167194] pinctrl core: registered pin 63 (io63) on rt2880-pinmux
[    0.167208] pinctrl core: registered pin 64 (io64) on rt2880-pinmux
[    0.167222] pinctrl core: registered pin 65 (io65) on rt2880-pinmux
[    0.167237] pinctrl core: registered pin 66 (io66) on rt2880-pinmux
[    0.167250] pinctrl core: registered pin 67 (io67) on rt2880-pinmux
[    0.167263] pinctrl core: registered pin 68 (io68) on rt2880-pinmux
[    0.167276] pinctrl core: registered pin 69 (io69) on rt2880-pinmux
[    0.167291] pinctrl core: registered pin 70 (io70) on rt2880-pinmux
[    0.167303] pinctrl core: registered pin 71 (io71) on rt2880-pinmux
[    0.167316] pinctrl core: registered pin 72 (io72) on rt2880-pinmux
[    0.167358] pinctrl core: add 2 pinctrl maps
[    0.167393] rt2880-pinmux pinctrl: found group selector 0 for i2c
[    0.167414] rt2880-pinmux pinctrl: found group selector 1 for uartf
[    0.167438] rt2880-pinmux pinctrl: request pin 1 (io1) for pinctrl
[    0.167457] rt2880-pinmux pinctrl: request pin 2 (io2) for pinctrl
[    0.167479] rt2880-pinmux pinctrl: request pin 7 (io7) for pinctrl
[    0.167498] rt2880-pinmux pinctrl: request pin 8 (io8) for pinctrl
[    0.167518] rt2880-pinmux pinctrl: request pin 9 (io9) for pinctrl
[    0.167537] rt2880-pinmux pinctrl: request pin 10 (io10) for pinctrl
[    0.167557] rt2880-pinmux pinctrl: request pin 11 (io11) for pinctrl
[    0.167577] rt2880-pinmux pinctrl: request pin 12 (io12) for pinctrl
[    0.167596] rt2880-pinmux pinctrl: request pin 13 (io13) for pinctrl
[    0.167615] rt2880-pinmux pinctrl: request pin 14 (io14) for pinctrl
[    0.167634] rt2880-pinmux pinctrl: failed to lookup the sleep state
[    0.182617] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.193642] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.207514] clocksource: Switched to clocksource systick
[    0.219443] NET: Registered protocol family 2
[    0.228927] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.242701] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.255291] TCP: Hash tables configured (established 1024 bind 1024)
[    0.268020] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.279545] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.292258] NET: Registered protocol family 1
[    0.300879] PCI: CLS 0 bytes, default 32
[    0.304613] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.317223] alarmtimer alarmtimer: no of_node; not parsing pinctrl DT
[    0.317559] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.329662] Crashlog allocated RAM at address 0x3f00000
[    0.355071] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.366593] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.388836] io scheduler noop registered
[    0.396487] io scheduler deadline registered (default)
[    0.406996] ralink-usb-phy usbphy: invalid resource
[    0.417253] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.430342] serial8250 serial8250: no of_node; not parsing pinctrl DT
[    0.430695] pinctrl core: add 1 pinctrl maps
[    0.430736] rt2880-pinmux pinctrl: found group selector 3 for uartlite
[    0.430763] rt2880-pinmux pinctrl: request pin 15 (io15) for 10000c00.uartlite
[    0.430784] rt2880-pinmux pinctrl: request pin 16 (io16) for 10000c00.uartlite
[    0.431041] console [ttyS0] disabled
[    0.438082] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    0.457754] console [ttyS0] enabled
[    0.471573] bootconsole [early0] disabled
[    0.488986] pinctrl core: add 1 pinctrl maps
[    0.489033] rt2880-pinmux pinctrl: found group selector 2 for spi
[    0.489062] rt2880-pinmux pinctrl: request pin 3 (io3) for 10000b00.spi
[    0.489083] rt2880-pinmux pinctrl: request pin 4 (io4) for 10000b00.spi
[    0.489102] rt2880-pinmux pinctrl: request pin 5 (io5) for 10000b00.spi
[    0.489122] rt2880-pinmux pinctrl: request pin 6 (io6) for 10000b00.spi
[    0.492282] spi spi0.0: force spi mode3
[    0.500519] m25p80 spi0.0: mx25l12805d (16384 Kbytes)
[    0.510675] 8 ofpart partitions found on MTD device spi0.0
[    0.521606] Creating 8 MTD partitions on "spi0.0":
[    0.531160] 0x000000000000-0x000000030000 : "u-boot"
[    0.542844] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.555313] 0x000000040000-0x000000050000 : "factory"
[    0.567276] 0x000000050000-0x000000df0000 : "firmware"
[    0.651836] 2 uimage-fw partitions found on MTD device firmware
[    0.663669] 0x000000050000-0x000000183be2 : "kernel"
[    0.675225] 0x000000183be2-0x000000df0000 : "rootfs"
[    0.687006] mtd: device 5 (rootfs) set to be root filesystem
[    0.698491] 1 squashfs-split partitions found on MTD device rootfs
[    0.710824] 0x0000004e0000-0x000000df0000 : "rootfs_data"
[    0.723443] 0x000000df0000-0x000000e00000 : "config"
[    0.735216] 0x000000e00000-0x000001000000 : "storage"
[    0.747139] 0x000000050000-0x000001000000 : "backup"
[    0.758953] 0x000000000000-0x000001000000 : "full"
[    0.771372] pinctrl core: add 1 pinctrl maps
[    0.771412] pinctrl core: add 1 pinctrl maps
[    0.771459] rt2880-pinmux pinctrl: found group selector 10 for rgmii2
[    0.771481] rt2880-pinmux pinctrl: found group selector 5 for mdio
[    0.771507] rt2880-pinmux pinctrl: request pin 60 (io60) for 10100000.ethernet
[    0.771529] rt2880-pinmux pinctrl: request pin 61 (io61) for 10100000.ethernet
[    0.771551] rt2880-pinmux pinctrl: request pin 62 (io62) for 10100000.ethernet
[    0.771572] rt2880-pinmux pinctrl: request pin 63 (io63) for 10100000.ethernet
[    0.771593] rt2880-pinmux pinctrl: request pin 64 (io64) for 10100000.ethernet
[    0.771614] rt2880-pinmux pinctrl: request pin 65 (io65) for 10100000.ethernet
[    0.771635] rt2880-pinmux pinctrl: request pin 66 (io66) for 10100000.ethernet
[    0.771655] rt2880-pinmux pinctrl: request pin 67 (io67) for 10100000.ethernet
[    0.771676] rt2880-pinmux pinctrl: request pin 68 (io68) for 10100000.ethernet
[    0.771696] rt2880-pinmux pinctrl: request pin 69 (io69) for 10100000.ethernet
[    0.771717] rt2880-pinmux pinctrl: request pin 70 (io70) for 10100000.ethernet
[    0.771737] rt2880-pinmux pinctrl: request pin 71 (io71) for 10100000.ethernet
[    0.771764] rt2880-pinmux pinctrl: request pin 22 (io22) for 10100000.ethernet
[    0.771785] rt2880-pinmux pinctrl: request pin 23 (io23) for 10100000.ethernet
[    0.774027] libphy: mdio: probed
[    0.780639] mtk_soc_eth 10100000.ethernet: using fixed link parameters
[    0.793718] mtk_soc_eth 10100000.ethernet: loaded mt7620 driver
[    0.806306] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    0.823396] rt2880_wdt 10000120.watchdog: Initialized
[    0.835068] NET: Registered protocol family 10
[    0.848461] NET: Registered protocol family 17
[    0.857400] bridge: automatic filtering via arp/ip/ip6tables has been deprecated. Update your scripts to load br_netfilter if you need this.
[    0.882593] 8021q: 802.1Q VLAN Support v1.8
[    0.908312] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    0.923315] Freeing unused kernel memory: 168K (803a6000 - 803d0000)
[    3.474431] init: Console is alive
[    3.481564] init: - watchdog -
[    6.260814] usbcore: registered new interface driver usbfs
[    6.271929] usbcore: registered new interface driver hub
[    6.282660] usbcore: registered new device driver usb
[    6.321702] rt2880-pinmux pinctrl: request pin 1 (io1) for pio:1
[    6.321746] rt2880-pinmux pinctrl: request pin 2 (io2) for pio:2
[    6.321771] rt2880-pinmux pinctrl: request pin 9 (io9) for pio:9
[    6.331675] SCSI subsystem initialized
[    6.345677] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    6.360353] ehci-platform: EHCI generic platform driver
[    6.381158] phy phy-usbphy.0: remote usb device wakeup disabled
[    6.392962] phy phy-usbphy.0: UTMI 16bit 30MHz
[    6.401843] ehci-platform 101c0000.ehci: EHCI Host Controller
[    6.413332] ehci-platform 101c0000.ehci: new USB bus registered, assigned bus number 1
[    6.429240] ehci-platform 101c0000.ehci: irq 26, io mem 0x101c0000
[    6.447186] ehci-platform 101c0000.ehci: USB 2.0 started, EHCI 1.00
[    6.460794] hub 1-0:1.0: no of_node; not parsing pinctrl DT
[    6.460851] hub 1-0:1.0: USB hub found
[    6.468741] hub 1-0:1.0: 1 port detected
[    6.479923] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    6.493800] ohci-platform: OHCI generic platform driver
[    6.504560] ohci-platform 101c1000.ohci: Generic Platform OHCI controller
[    6.518162] ohci-platform 101c1000.ohci: new USB bus registered, assigned bus number 2
[    6.534037] ohci-platform 101c1000.ohci: irq 26, io mem 0x101c1000
[    6.568763] hub 2-0:1.0: no of_node; not parsing pinctrl DT
[    6.568837] hub 2-0:1.0: USB hub found
[    6.576780] hub 2-0:1.0: 1 port detected
[    6.586501] rt2880-pinmux pinctrl: request pin 7 (io7) for pio:7
[    6.586737] rt2880-pinmux pinctrl: request pin 8 (io8) for pio:8
[    6.586882] rt2880-pinmux pinctrl: request pin 10 (io10) for pio:10
[    6.587023] rt2880-pinmux pinctrl: request pin 11 (io11) for pio:11
[    6.587166] rt2880-pinmux pinctrl: request pin 14 (io14) for pio:14
[    6.592617] usbcore: registered new interface driver usb-storage
[    6.609051] init: - preinit -
[    7.656574] mtk_soc_eth 10100000.ethernet eth0: port 4 link up (1000Mbps/Full duplex)
[    7.672941] 8021q: adding VLAN 0 to HW filter on device eth0
[    7.840751] random: procd: uninitialized urandom read (4 bytes read, 12 bits of entropy available)
[    8.753573] mount_root: loading kmods from internal overlay
[   10.112899] jffs2_scan_eraseblock(): End of filesystem marker found at 0x10000
[   10.127354] jffs2_build_filesystem(): unlocking the mtd device... done.
[   10.140541] jffs2_build_filesystem(): erasing all blocks after the end marker... done.
[   63.656805] jffs2: notice: (379) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   63.688065] block: attempting to load /tmp/jffs_cfg/upper/etc/config/fstab
[   63.702033] block: unable to load configuration (fstab: Entry not found)
[   63.715525] block: attempting to load /tmp/jffs_cfg/etc/config/fstab
[   63.728369] block: unable to load configuration (fstab: Entry not found)
[   63.742673] block: attempting to load /etc/config/fstab
[   63.810395] block: unable to load configuration (fstab: Entry not found)
[   63.823945] block: no usable configuration
[   64.023556] jffs2: notice: (375) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   65.154974] block: attempting to load /tmp/jffs_cfg/upper/etc/config/fstab
[   65.168980] block: unable to load configuration (fstab: Entry not found)
[   65.182498] block: attempting to load /tmp/jffs_cfg/etc/config/fstab
[   65.195353] block: unable to load configuration (fstab: Entry not found)
[   65.208856] block: attempting to load /etc/config/fstab
[   65.219441] block: unable to load configuration (fstab: Entry not found)
[   65.232925] block: no usable configuration
[   65.242641] mount_root: overlay filesystem has not been fully initialized yet
[   65.259772] mount_root: switching to jffs2 overlay
[   66.439466] urandom-seed: Seed file not found (/etc/urandom.seed)
[   66.683470] mtk_soc_eth 10100000.ethernet eth0: port 4 link down
[   66.705305] procd: - early -
[   66.711142] procd: - watchdog -
[   67.356226] procd: - ubus -
[   67.385976] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.404561] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.422722] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.441226] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.459375] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.477659] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.495955] random: ubusd: uninitialized urandom read (4 bytes read, 110 bits of entropy available)
[   67.514460] procd: - init -
[   67.835153] tun: Universal TUN/TAP device driver, 1.6
[   67.845291] tun: (C) 1999-2004 Max Krasnyansky <maxk@qualcomm.com>
[   67.863424] ip6_tables: (C) 2000-2006 Netfilter Core Team
[   67.889374] rtl8367b rtl8367rb: gpios missing in devictree
[   67.904954] Loading modules backported from Linux version wt-2016-10-03-1-g6fcb1a6
[   67.920111] Backport generated by backports.git backports-20160324-9-g0e38f5c
[   67.998483] ip_tables: (C) 2000-2006 Netfilter Core Team
[   68.022280] nf_conntrack version 0.5.0 (1967 buckets, 7868 max)
[   68.086335] xt_time: kernel timezone is -0000
[   68.108565] PPP generic driver version 2.4.2
[   68.120018] NET: Registered protocol family 24
[   68.148653] rt2800_wmac 10180000.wmac: loaded eeprom from mtd device "factory"
[   68.163123] ieee80211 phy0: rt2x00_set_rt: Info - RT chipset 5390, rev 0500 detected
[   68.178574] ieee80211 phy0: rt2x00_set_rf: Info - RF chipset 7620 detected
[   68.192994] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   69.033293] random: jshn: uninitialized urandom read (4 bytes read, 113 bits of entropy available)
[   69.123032] random: jshn: uninitialized urandom read (4 bytes read, 113 bits of entropy available)
[   77.190027] mtk_soc_eth 10100000.ethernet eth0: port 4 link up (1000Mbps/Full duplex)
[   77.205991] 8021q: adding VLAN 0 to HW filter on device eth0
[   77.285719] device eth0.1 entered promiscuous mode
[   77.295358] device eth0 entered promiscuous mode
[   77.348243] br-lan: port 1(eth0.1) entered forwarding state
[   77.359453] br-lan: port 1(eth0.1) entered forwarding state
[   78.776975] random: nonblocking pool is initialized
[   79.356937] br-lan: port 1(eth0.1) entered forwarding state
[   82.524490] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   82.541170] device wlan0 entered promiscuous mode
[   82.550704] br-lan: port 2(wlan0) entered forwarding state
[   82.561695] br-lan: port 2(wlan0) entered forwarding state
[   82.605918] br-lan: port 2(wlan0) entered disabled state
[   87.406056] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   87.418951] br-lan: port 2(wlan0) entered forwarding state
[   87.429955] br-lan: port 2(wlan0) entered forwarding state
[   88.390666] br-lan: port 2(wlan0) entered forwarding state`

mkresin commented:
While I was about to merge this one, I've spotted more stuff.

But before we proceed, please remove the  "rtl8367b: add direct mdio bus support" commit. I will not accept any new features. Please create a separate PR for this one.

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+#include "mt7620a.dtsi"
+
+/ {
+	compatible = "ralink,mt7620a-eval-board", "ralink,mt7620a-soc";
```
drop this line

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+			reg = <0x50000 0xda0000>;
+		};
+
+		partition@df0000 {
```
why do you keep the config, storage and backup partitions? The are only relevant for the stock firmware and not for LEDE.

13hakta replied:
Sure, they are. Thought it would be used somehow. May be dropped. Also "full" partition with them.

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+			reg = <0xe00000 0x200000>;
+		};
+
+		backup@50000 {
```
partition@

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+			read-only;
+		};
+
+		full@0 {
```
Why do yo have a partition that covers the whole flash?

13hakta replied:
Legacy from stock firmware. May be dropped taking in account mentioned above.

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+	status = "okay";
+};
+
+&pcie {
```
This is already disabled in mt7620a.dtsi.

mkresin commented on target/linux/ramips/dts/kng_rc.dts:
```diff
@@ -0,0 +1,187 @@
+#include <dt-bindings/input/input.h>
```
move the includes after ``#include "mt7620a.dtsi"`` and order them alphabetical.

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -37,6 +37,10 @@ define Build/elecom-header
 endef
 
+define Build/zyimage
+	zyimage $(1) $@
```
$(STAGING_DIR_HOST)/bin/zyimage $(1) $@

Otherwise it might runs a zyimage from you system instead the LEDE one.

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -418,3 +422,13 @@ define Device/dch-m225
+  DEVICE_TITLE := ZyXEL Keenetic Viva
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport kmod-switch-rtl8366-smi kmod-switch-rtl8367b
+  IMAGES += factory.bin
+  IMAGE/factory.bin := $$(IMAGE/sysupgrade.bin) | zyimage -d 8997 -v "ZyXEL Keenetic Viva LEDE"
```
"ZyXEL Keenetic Viva"

mkresin commented:
finally merged
